### ISSUE TITLE
Improve flow sensitivity, using SSA form.

### ISSUE THREAD

**Issue by markshannon** [python/mypy#16594]
**Feature**

Improve local flow sensitivity of the analysis to reduce the number of false positives, using SSA form.

While MyPy has some flow sensitivity, it is can be tripped up by relatively simple flow control.
Using a better internal representation of variables and dataflow would allow better type checking without adding complexity to the rest of MyPy.

**Pitch**

For example:
```Py
def foo(a: list[str], b: list[int]) -> None:
      x = 0
      for i in a:
           x += i
           print(x)
      x = ""
      for i in b:
           x += i
           print(x)
```
The above function is clearly safe, but MyPy complains.
With SSA form, MyPy could trivially determine that the function is corrrect by treating `x` and `i` as two distinct SSA variables 
```Py
def foo(a: list[str], b: list[int]) -> None:
      x_0 = 0
      for i_0 in a:
           x_0 += i_0
           print(x_0)
      x_1 = ""
      for i_1 in b:
           x_1 += i_1
           print(x_1)
```

MyPy already supports union types, so phi-nodes can be straightforwardly typed.

**Comment by hauntsaninja** [comment]
This is fixed with `--allow-redefinition-new`
